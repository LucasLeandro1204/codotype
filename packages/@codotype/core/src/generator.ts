import { ConfigurationGroup } from "./configuration-option-types";
import { Datatype } from "./datatype";
import { RelationType, Relation } from "./relation";
import { Schema } from "./schema";
import { Attribute } from "./attribute";
import { AttributeAddon } from "./attribute-addon";
import { Project } from "./project";

export enum ExperienceRecommendation {
  BEGINNER = "beginner",
  JUNIOR = "junior",
  INTERMEDIATE = "intermediate",
  EXPERT = "expert",
}

// QUESTION - should the be split up into better separation of concerns? i.e. AttributeEditorConfiguration, SchemaEditorConfiguration, RelationEditorConfiguration
export interface SchemaEditorConfiguration {
  documentation: string; // Any documentation for the SchemaEditor
  configurationGroups: ConfigurationGroup[]; // ConfigurationGroup
  defaultSchemas: Schema[]; // QUESTION - how do we enforce Schema.source = SchemaSource.GENERATOR?
  supportedDatatypes: Datatype[]; // The datatypes supported by this generator
  supportedRelations: RelationType[]; // The relation types supported by this generator
  defaultAttributes: Attribute[]; // Default attributes applied to _every_ new Attribute - QUESTION - how do we enforce Addon value for any Attributes defined here?
  attributeAddons: AttributeAddon[]; // Addons made available to the AttributeEditor
  enableAttributeDefaultValue: boolean; // Whether or not to enable the `Default Value` input in the AttributeEditor
  defaultRelations: Relation[];
}

/**
 * GeneratorCreatorContactOption
 * Provides enum of available types of contact information for a GeneratorCreator
 */
enum GeneratorCreatorContactOption {
  github = "github",
  gitlab = "gitlab",
  twitter = "twitter",
  linkedin = "linkedin",
  email = "email",
  website = "website",
}

/**
 * GeneratorCreator
 * Stores metadata about the creator of the Generator
 */
export interface GeneratorCreator {
  name: string;
  contact: {
    [key in GeneratorCreatorContactOption]?: string;
  };
}

// TODO - GeneratorMeta should be split up into:
// Generator
// GeneratorMeta
// Generator encapsulate the GeneratorMeta + link to the entry point of the generator
// Can be defined in `index.ts` in the root of the generator, or split up into smaller components in `src`
// TODO - update this to include example projects (must be bundled at the generator-level)
// TODO - add `defaultProjectName` property, optional
export interface GeneratorMeta {
  id: string; // unique ID for the generator
  label: string; // short label for the generator
  description: string; // brief description of the generator
  documentation: string; // More thorough documentation of the Generator
  icon: string; // URL to the generator's icon. Must be at least 200x200px
  homepage: string; // the "homepage" for this generator
  version: string; // the current version of the generator
  codotypeVersion: string; // the version of codotype that runs this generator
  createdBy: GeneratorCreator; // Metadata about the
  techTags: string[]; // an array of strings describing the tech used in the generator
  typeTags: string[]; // describes the type of codebase produced by this generator
  experience: ExperienceRecommendation; // an optional tag detailing the level of experience required to use the code produced by the generator
  project_path: string; // the name of the directory for the generator's output
  schemaEditorConfiguration: SchemaEditorConfiguration;
  configurationGroups: ConfigurationGroup[]; // an array of OptionGroup objects that expose additional configuration provided by the generator
  exampleProjects: Project[];
}
