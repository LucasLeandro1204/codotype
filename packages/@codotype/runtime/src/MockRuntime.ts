import {
    PluginRegistration,
    RuntimeLogLevel,
    Datatype,
    RelationType,
    RuntimeAdaptor,
    RuntimeConstructorParams,
    Runtime,
} from "@codotype/core";
import * as path from "path";
import * as ejs from "ejs";
import { logger } from "./utils/logger";

// // // //

// Mock CodotypeRuntime class definition
export class MockRuntime implements Runtime {
    options: RuntimeConstructorParams;
    plugins: PluginRegistration[];

    _mocks_: {
        ensuredDir: string;
        copiedDirSrc: string;
        copiedDirDest: string;
        files: { [key: string]: string };
    };

    // constructor
    // Handles options to run a single generator instance
    constructor(options: RuntimeConstructorParams) {
        // Assigns this.options
        this.options = options;

        // Sets default mocks
        this._mocks_ = {
            ensuredDir: "",
            copiedDirDest: "",
            copiedDirSrc: "",
            files: {},
        };

        // Assigns this.plugins
        this.plugins = [];

        // Assigns this.options.cwd
        this.options.cwd = process.cwd();

        // Returns the runtime instance
        return this;
    }

    /**
     * log
     * A logging function used internally by the Runtime
     * Invokes logger function with args, options, and Runtime.options.logLevel
     * @param args - see logger.ts
     */
    log(args: any, options: { level: RuntimeLogLevel }) {
        logger(args, options, this.options.logLevel);
    }

    registerPlugin(props: {
        modulePath?: string;
        relativePath?: string;
        absolutePath?: string;
    }): Promise<PluginRegistration> {
        return Promise.resolve(this.plugins[0]);
    }

    execute(): Promise<void> {
        return Promise.resolve();
    }

    // ensureDir
    // Ensures presence of directory for template compilation
    // TODO - this is repeated in @codotype/generator - should be abstracted, or only encapsulated in the runtime
    async ensureDir(dir: string): Promise<boolean> {
        this._mocks_.ensuredDir = dir;
        return new Promise((resolve, reject) => {
            return resolve(true);
        });
    }

    // templatePath
    templatePath(generatorResolved: string, template_path = "./") {
        // return path.join(generatorResolved, "templates", template_path);
        return path.join(generatorResolved, "templates", template_path);
    }

    // destinationPath
    destinationPath(destPath: string, dest = "./") {
        return path.join(destPath, dest);
    }

    // renderTemplate
    // Compiles an EJS template and returns the result
    renderTemplate(
        generatorInstance: RuntimeAdaptor,
        src: string,
        options = {},
    ) {
        return new Promise<string>((resolve, reject) => {
            // default options padded into the renderFile
            let renderOptions = {};

            // // // //
            // // // //
            // TODO - document this structure
            // The `data` object is passed into each file that gets rendered
            // TODO - this should be abstracted into a separate function
            const data = {
                project: generatorInstance.options.project,
                meta: generatorInstance.options.plugin,
                configuration: generatorInstance.options.project.configuration,
                helpers: {
                    // indent,
                    // trailingComma,
                },
                Datatype: Datatype,
                RelationType: RelationType,
                ...options, // QUESTION - are options ever used here?
            };
            // // // //
            // // // //

            // Compiles EJS template
            return ejs.renderFile(src, data, renderOptions, (err, str) => {
                // Handles template compilation error
                if (err) return reject(err);

                // Resolves with compiled template
                return resolve(str);
            });
        });
    }

    fileExists(dest: string): Promise<boolean> {
        // return this.fs.existsSync(dest);
        return Promise.resolve(true);
    }

    compareFile(dest: string, compiledTemplate: string): Promise<boolean> {
        return Promise.resolve(this._mocks_.files[dest] === compiledTemplate);
    }

    writeFile(dest: string, compiledTemplate: string) {
        this._mocks_ = {
            ...this._mocks_,
            files: {
                ...this._mocks_.files,
                [dest]: compiledTemplate,
            },
        };
        return Promise.resolve(true);
    }

    // copyDir
    copyDir(src: string, dest: string): Promise<boolean> {
        this._mocks_.copiedDirSrc = src;
        this._mocks_.copiedDirDest = dest;
        return new Promise((resolve, reject) => {
            return resolve(true);
        });
    }

    // composeWith
    async composeWith(
        parentGeneratorInstance: RuntimeAdaptor, // TODO - fix this
        generatorModule: string,
        options = {},
    ) {}
}
