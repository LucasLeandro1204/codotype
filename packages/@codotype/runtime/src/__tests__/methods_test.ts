import { MockRuntime } from "../MockRuntime";
import { CodotypeGenerator } from "../generator";
import {
    project,
    runtimeConstructorOptions,
    baseGeneratorOptions,
    generatorPrototype,
    generatorPrototype01,
    generatorPrototype02,
} from "./test_state";

// // // //

describe("templatePath method", () => {
    it("should define correct template path", () => {
        const dest = "destination";
        const template = "template.json";
        const resolved = "my/resolved/path";
        const runtime = new MockRuntime(runtimeConstructorOptions);

        const generatorOptions = {
            ...baseGeneratorOptions,
            runtime,
            dest,
            resolved,
        };
        const generatorInstance = new CodotypeGenerator(
            generatorPrototype,
            generatorOptions,
        );

        const templatePath = generatorInstance.templatePath(template);
        expect(templatePath).toBe(`${resolved}/templates/${template}`);
    });
});

describe("destinationPath method", () => {
    it("should define correct destination path", () => {
        const dest = "destination";
        const dirName = "testyMcTestface";
        const runtime = new MockRuntime(runtimeConstructorOptions);

        const generatorOptions = {
            ...baseGeneratorOptions,
            runtime,
            dest,
            resolved: __dirname,
        };

        // NOTE - this will ALWAYS be handled by the runtime
        // Perhaps this `CodotypeGenerator` should be called `RuntimeAdaptor`?
        // WHAT DOES IT DO? -> this takes the generator properties defined by the a plugin author
        // and links them to an active Codotype Runtime instance
        // WHY - this prevents plugin/generator authors from having to worry about the runtime altogether - the runtime adaptor "joins" a generator to
        // its runtime so it can produce files. Without the runtime, the generator is little more than an object.
        const generatorInstance = new CodotypeGenerator(
            generatorPrototype,
            generatorOptions,
        );

        const destinationPath = generatorInstance.destinationPath(dirName);
        expect(destinationPath).toBe(`${dest}/${dirName}`);
    });
});

describe("ensureDir method", () => {
    it("should properly define mockRuntime._mocks_.ensuredDir", () => {
        const dest = "destination";
        const dirName = "testyMcTestface";
        const runtime = new MockRuntime(runtimeConstructorOptions);

        const generatorOptions = {
            ...baseGeneratorOptions,
            runtime,
            dest,
            resolved: __dirname,
        };
        const generatorInstance = new CodotypeGenerator(
            generatorPrototype,
            generatorOptions,
        );

        generatorInstance.ensureDir(dirName).then(() => {
            expect(runtime._mocks_.ensuredDir).toBe(`${dest}/${dirName}`);
        });
    });
});

describe("copyDir method", () => {
    it("should properly define copiedDirSrc and copiedDirDest on mockRuntime._mocks_", () => {
        const src = "mySrc";
        const dest = "myDest";
        const runtime = new MockRuntime(runtimeConstructorOptions);

        const generatorOptions = {
            ...baseGeneratorOptions,
            runtime,
            dest,
            resolved: "dir",
        };

        const generatorInstance = new CodotypeGenerator(
            generatorPrototype,
            generatorOptions,
        );

        generatorInstance.copyDir({ src, dest }).then(() => {
            expect(runtime._mocks_.copiedDirSrc).toBe("dir/templates/mySrc");
            expect(runtime._mocks_.copiedDirDest).toBe("myDest/myDest"); // TODO - why is it like that?
        });
    });
});

describe("renderComponent", () => {
    test("renders", async () => {
        const dest = "destination";
        const runtime = new MockRuntime(runtimeConstructorOptions);

        // // // //
        // NOTE - all of this is handled by the runtime -> should this be handled inside the RuntimeGeneratorAdaptor?
        const generatorOptions = {
            ...baseGeneratorOptions,
            runtime,
            dest,
            resolved: __dirname, // NOTE - need to use __dirname here beacuse the `templates` directory sits next to this test
        };

        const generatorInstance = new CodotypeGenerator(
            generatorPrototype01,
            generatorOptions,
        );

        // Tests write function
        await generatorInstance.write({ project, runtime: generatorInstance });

        // @ts-ignore
        await generatorInstance.forEachSchema({
            project,
            schema: project.schemas[1],
            runtime: generatorInstance,
        });

        // @ts-ignore
        await generatorInstance.forEachSchema({
            project,
            schema: project.schemas[0],
            runtime: generatorInstance,
        });
        //
        // // // //

        // console.log(runtime._mocks_);
        expect(runtime._mocks_.files).toMatchSnapshot();
    });
});

// // // //
// OLD CONSTRUCTOR TESTS

// const CodotypeGenerator = require('../../lib/generator.js')
// const chai = require("chai");
// const assert = chai.assert;

// // // // //

// describe('@codotype/generator', () => {

//   describe('constructor', () => {

//     it('throw error if options.runtime is undefined', () => {
//       const generatorPrototype = {}
//       const generatorOptions = {}

//       try {
//         new CodotypeGenerator(generatorPrototype, generatorOptions)
//       } catch (err) {
//         assert.equal(true, true)
//       }
//     });

//     it('throw error if options.resolved is undefined', () => {
//       const generatorPrototype = {}
//       const generatorOptions = { runtime: true }

//       try {
//         new CodotypeGenerator(generatorPrototype, generatorOptions)
//       } catch (err) {
//         assert.equal(true, true)
//       }
//     });

//     it('throw error if generatorPrototype is missing write, writeSchema, and compileInPlace properties ', () => {
//       const generatorPrototype = {}
//       const generatorOptions = { runtime: true, resolved: true }

//       try {
//         new CodotypeGenerator(generatorPrototype, generatorOptions)
//       } catch (err) {
//         assert.equal(true, true)
//       }
//     });

//   });

// });
