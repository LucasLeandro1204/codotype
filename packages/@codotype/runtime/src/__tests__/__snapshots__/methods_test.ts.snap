// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderComponent renders 1`] = `
Object {
  "": "{
name: \\"project\\"
}",
}
`;
